### ISSUE TITLE
[suggestion] Extend Password Type

### ISSUE THREAD

**Issue by tylkomat** [keystonejs/keystone#1339]
Keystone should offer a more fine grained handling for passwords. Now when changing a password the user just has to type the new password twice. That is fine if your logged in user is an admin and you edit some other user. If you edit your own user, there should be a different handling. It is more secure, if the user has to enter the old password again instead of the new password repetition.

I someone somehow takes over your session, he can not change your password.

**Comment by morenoh149** [comment]
the current flow has a user respond to an email no? and the scenario is a user the forgot the password no?

I think we'd be more likely to make such a change if you linked to an OWASP paper or page detailing a secure flow.

**Comment by tylkomat** [comment]
No I was talking about the general behaviour of the password type. In `updateHandler.js` when updating a password. The handler expects that there are 2 times the same passwords submitted. This leaves not much flexibility to the developer. 

For the reset password path this is fine. Click the email link -> enter your new password twice (to just not forget it again too fast, although once would be fine too in this case). 
But when you login and want to change your own password I think it is better to confirm with your old password that you want to set a new password. It should be obvious to see that this is more secure compared to the other case where you don't have to remember your old password.
When a admin changes the password of another user he should also confirm with his own password (not the users password, because he doesn't know it)

The Example was before that someone somehow takes over your session. So he's logged in without using your password. He can't take over your account by locking you out, when he changes your password, because he doesn't know your old password.

Just a side note, changing the email address of a user should be also confirmed by the password of the user or if it was changes by an admin with the password of the admin. Otherwise an attacker could change the email and then use the forgot password function to retrieve a new password.

I know that the password recovery is still in development. But might be good to keep in mind.

**Comment by creynders** [comment]
@tylkomat I agree. Completely. There's a huge number of improvements that need to be made and this is definitely one of those. That said, we need to distribute our efforts and concentration and prioiritise things, and _to me_ this doesn't make top of the list (yet). But we'll definitely keep this in mind.
